cli/runtime/config/config_factory.go

```diff
@@ -8,22 +8,184 @@ import (
+	}
+
+	return makeMultiFileCfg(cfgNode, cfgNextGenNode)
 }
 
-type CfgOpts func(config *CfgOptions)
+// Construct multi file node from config.yaml and config-ng.yaml nodes
+func makeMultiFileCfg(cfgNode, cfgNextGenNode *yaml.Node) (*yaml.Node, error) {
+	// config items that goes to config-ng.yaml
+	nextGenCfgItems := []string{
+		KeyContexts,
+		KeyCurrentContext,
+	}
+
+	// Process the next gen items and discard both key node and value node from config.yaml
+	for _, nextGenItem := range nextGenCfgItems {
+		// Find the next gen item node from the config.yaml
+		nextGenItemNodeIndex := nodeutils.GetNodeIndex(cfgNode.Content[0].Content, nextGenItem)
+		if nextGenItemNodeIndex == -1 {
+			continue
+		}
+		// Delete the next gen item node key from the config.yaml
+		cfgNode.Content[0].Content = append(cfgNode.Content[0].Content[:nextGenItemNodeIndex-1], cfgNode.Content[0].Content[nextGenItemNodeIndex:]...)
+		// Delete the next gen item node value from the config.yaml
+		cfgNode.Content[0].Content = append(cfgNode.Content[0].Content[:nextGenItemNodeIndex-1], cfgNode.Content[0].Content[nextGenItemNodeIndex:]...)
+	}
+
+	// Construct Multi Bytes data
+	var multiBytes []byte
+	var multiNode yaml.Node
+
+	// construct cfgNode bytes and append to multiBytes
+	multiBytes, err := nodeutils.AppendNodeBytes(multiBytes, cfgNode)
+	if err != nil {
+		return nil, err
+	}
+
+	// construct cfgNextGenNode bytes and append to multiBytes
+	multiBytes, err = nodeutils.AppendNodeBytes(multiBytes, cfgNextGenNode)
+	if err != nil {
+		return nil, err
+	}
 
-func WithCfgPath(path string) CfgOpts {
-	return func(config *CfgOptions) {
-		config.CfgPath = path
+	// create new yaml node if multiBytes contains no data or empty
+	if multiBytes == nil {
+		multiNode, err := newClientConfigNode()
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to create new client config node")
+		}
+
+		return multiNode, err
 	}
+
+	// construct the multi node from multi bytes data
+	if len(multiBytes) != 0 {
+		err := yaml.Unmarshal(multiBytes, &multiNode)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &multiNode, nil
+}
+
+// persistConfig write the updated node data to config.yaml and config-ng.yaml based on few
+func persistConfig(node *yaml.Node) error {
+	useUnifiedConfig, err := UseUnifiedConfig()
+	if err != nil {
+		useUnifiedConfig = false
+	}
+	// If useUnifiedConfig is set to true write to config-ng.yaml
+	if useUnifiedConfig {
+		return persistClientConfigNextGen(node)
+	}
+
+	cfgNode, err := getClientConfigNoLock()
+	if err != nil {
+		return err
+	}
+
+	// deep copy of change node
+	var cfgNodeToPersist yaml.Node
+	data, err := yaml.Marshal(node)
+	if err != nil {
+		return err
+	}
+	err = yaml.Unmarshal(data, &cfgNodeToPersist)
+	if err != nil {
+		return err
+	}
+
+	cfgNextGenNode, err := getClientConfigNextGenNodeNoLock()
+	if err != nil {
+		return err
+	}
+
+	// next gen config items that goes to config-ng.yaml
+	nextGenItemNodes := []*yaml.Node{
```

Comment:
something for consideration:

since the set keys of nodes we intent to write back to the legacy config file is not only known, but fixed _forever_, 
would it be cleaner to hard code that list instead? We can then loop through the keys of the incoming node's context and crosscheck against that list, and update the appropriate config/configng node. 

The implementation would then not need updating when new config sections are added to config-ng, and we run no risk of somehow introducing new stuff back into the legacy config.
We also avoid need to do the content resetting bit in L164

Reply:
Changed the Implementation it is now much simpler